pkg/kp/dsstore/consul_store.go

```diff
@@ -0,0 +1,207 @@
+package dsstore
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/pborman/uuid"
+	klabels "k8s.io/kubernetes/pkg/labels"
+
+	"github.com/square/p2/pkg/ds/fields"
+	"github.com/square/p2/pkg/kp/consulutil"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/types"
+	"github.com/square/p2/pkg/util"
+)
+
+type kvPair struct {
+	key   string
+	value []byte
+}
+
+type consulKV interface {
+	Get(key string, opts *api.QueryOptions) (*api.KVPair, *api.QueryMeta, error)
+	List(prefix string, opts *api.QueryOptions) (api.KVPairs, *api.QueryMeta, error)
+	CAS(pair *api.KVPair, opts *api.WriteOptions) (bool, *api.WriteMeta, error)
+}
+
+type consulStore struct {
+	applicator labels.Applicator
+	kv         consulKV
+	logger     logging.Logger
+	retries    int
+}
+
+// TODO: combine with similar CASError type in pkg/labels
+type CASError string
+
+func (e CASError) Error() string {
+	return fmt.Sprintf("Could not check-and-set key %q", string(e))
+}
+
+// Make sure functions declared in the Store interface have the proper contract
+var _ Store = &consulStore{}
+
+func NewConsul(client *api.Client, retries int, logger *logging.Logger) *consulStore {
```

Comment:
usually I return the interface type (e.g. `Store`) from these, but since it's an unexported type i'm not sure what will happen?